### PR TITLE
first commit of LFHCAL benchmarks

### PR DESCRIPTION
**Pull request by steinber**
This is based on the zdc_lyso examples but should work once I iron out a few things.  I've cleared with Dima that I'll work out those things in the PR discussion.

### Briefly, what does this PR introduce?

Benchmarks for LFHCAL, using a Snakemake workflow based on zdc_lyso example.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ X] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ X] Tests for the changes have been added 
- [ X] Documentation has been added / updated
- [ X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
None

### Does this PR change default behavior?
It adds the LFHCAL benchmarks to detector_benchmarks, so yes it may have an impact on the centralized benchmark production.

### CODE REVIEW

**Comment by veprbl** [comment]
Another place that needs an entry is
https://github.com/eic/detector_benchmarks/blob/a40a48bbfc0e0db6932cd22d96b5219cfc7879c2/.gitlab-ci.yml#L135-L147

File: benchmarks/lfhcal/doCompare_widebins_mom.C [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  
+  mgMom->Add(gr1);
+  c_mom->cd();
+  mgMom->GetXaxis()->SetRangeUser(0.40,20.2);
+  mgMom->GetYaxis()->SetRangeUser(0.0,1.50*TMath::MaxElement(gr1->GetN(),gr1->GetY())); // 50% more of the maximum value on yaxis
+  mgMom->Draw("AP");
+  lmom->AddEntry(gr1,"Nominal");
+  lmom->Draw("same");
+  //draw_req_Mom(etamin,etamax,0.,mgMom->GetXaxis()->GetXmax());
+  c_mom->SaveAs(Form("Final_Results/%s/mom/lfhcal_mom_resol_%1.1f_eta_%1.1f.png",particle.Data(),etamin,etamax));
```
Comment (veprbl) [inline benchmarks/lfhcal/doCompare_widebins_mom.C:122]:
```suggestion
  c_mom->SaveAs(Form("Final_Results/%s/mom/lfhcal_mom_resol_%s_%1.1f_eta_%1.1f.png",particle.Data(),particle.Data(),etamin,etamax));
```

File: benchmarks/lfhcal/Snakefile [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            [
+                "{{CAMPAIGN}}/{PARTICLE}/mom/lfhcal_mom_{MOMENTUM:.1f}_mom_resol_{PARTICLE}.root",
+            ],
+            MOMENTUM=[0.5, 1.0, 2.0, 5.0, 10.0, 20.0],
+	    PARTICLE=["neutron", "pi-"]
+        ),
+        script="benchmarks/lfhcal/doCompare_widebins_mom.C",
+    output:
+        "{CAMPAIGN}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{ETA_MIN}_eta_{ETA_MAX}.png",
+        "{CAMPAIGN}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{ETA_MIN}_eta_{ETA_MAX}.root",
```
Comment (veprbl) [inline benchmarks/lfhcal/Snakefile:88]:
```suggestion
                "{{CAMPAIGN}}/{{PARTICLE}}/mom/lfhcal_mom_{MOMENTUM:.1f}_mom_resol_{{PARTICLE}}.root",
            ],
            MOMENTUM=[0.5, 1.0, 2.0, 5.0, 10.0, 20.0],
        ),
        script="benchmarks/lfhcal/doCompare_widebins_mom.C",
    output:
        "{CAMPAIGN}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{PARTICLE}_{ETA_MIN}_eta_{ETA_MAX}.png",
        "{CAMPAIGN}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{PARTICLE}_{ETA_MIN}_eta_{ETA_MAX}.root",
```

File: benchmarks/lfhcal/Snakefile [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+"""
+
+LFHCAL_ETA_BINS = [1.2,1.5,2,2.5,3,3.5]
+
+rule lfhcal:
+    input:
+        expand(
+            [
+	      "{{CAMPAIGN}}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{ETA_BIN}.png",
+	      "{{CAMPAIGN}}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{ETA_BIN}.root",
```
Comment (veprbl) [inline benchmarks/lfhcal/Snakefile:116]:
```suggestion
	      "{{CAMPAIGN}}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{PARTICLE}_{ETA_BIN}.png",
	      "{{CAMPAIGN}}/Final_Results/{PARTICLE}/mom/lfhcal_mom_resol_{PARTICLE}_{ETA_BIN}.root",
```

File: benchmarks/lfhcal/doCompare_widebins_mom.C [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  
+  for (int i=0; i<size; i++){
+    p[i] = momV.at(i);
+    sigma_p[i] = momresolV.at(i);
+    err_sigma_p[i] = err_momresolV.at(i);
+    err_p[i] = 0.;
+  }
+  
+  
+  TFile *fout = new TFile(Form("Final_Results/%s/mom/lfhcal_mom_resol_%1.1f_eta_%1.1f.root",particle.Data(),etamin,etamax),"recreate");
```
Comment (veprbl) [inline benchmarks/lfhcal/doCompare_widebins_mom.C:103]:
```suggestion
  TFile *fout = new TFile(Form("Final_Results/%s/mom/lfhcal_mom_resol_%s_%1.1f_eta_%1.1f.root",particle.Data(),particle.Data(),etamin,etamax),"recreate");
```

**Comment by steinber** [comment]
I just updated the branch - can it be merged when the checks pass?

**Comment by veprbl** [comment]
We first need to revert 64ae88b19b9b70c0cf75c8a8eeadc63f4640a933 and ffdfae931e8a8cd218bb651f310a21bf5fa6fa1f

**Comment by steinber** [comment]
Oh I see - this was for testing on the CI side.  I'm trying again wit a fresh branch checkout to see if the plots are empty in that case.

**Comment by steinber** [comment]
Hi @veprbl - I just ran straight from the branch (snakemake --cores 4 lfhcal_local) with 100 events and all plots were filled.  Not sure how to square what you're seeing and what I am seeing...

**Comment by veprbl** [comment]
> Hi @veprbl - I just ran straight from the branch (snakemake --cores 4 lfhcal_local) with 100 events and all plots were filled. Not sure how to square what you're seeing and what I am seeing...

Are you running everyting with Snakemake? Which version of the geometry do you source?

**Comment by veprbl** [comment]
> Oh I see - this was for testing on the CI side. I'm trying again wit a fresh branch checkout to see if the plots are empty in that case.

That should not matter for local testing.

**Comment by steinber** [comment]
I was using epic-main, since I assumed it was "nominal" - not so?

**Comment by veprbl** [comment]
detector_benchmarks also
https://github.com/eic/detector_benchmarks/blob/8116047447bcfa6aa493ea3dd4f6ab3e724b5c7f/.gitlab-ci.yml#L111

**Comment by steinber** [comment]
So the checks failed, but it's not an LFHCAL problem, but "ecal_gaps" raising a divide-by-zero error.

**Comment by steinber** [comment]
Something is very weird - when I go into the "sim:lfhcal: [neutron, 20GeV]", and check the log, the simulation is all 500 MeV electrons.  This may be part of the problem.

**Comment by veprbl** [comment]
> So the checks failed, but it's not an LFHCAL problem, but "ecal_gaps" raising a divide-by-zero error.

Correct. That might be a relevant issue.



> Something is very weird - when I go into the "sim:lfhcal: [neutron, 20GeV]", and check the log, the simulation is all 500 MeV electrons. This may be part of the problem.

Interesting. I will try to disable download cache, in case it causes wrong steering files to be retrieved from it.

**Comment by steinber** [comment]
> > So the checks failed, but it's not an LFHCAL problem, but "ecal_gaps" raising a divide-by-zero error.
> 
> Correct. That might be a relevant issue.

Relevant to ECAL :)

> 
> > Something is very weird - when I go into the "sim:lfhcal: [neutron, 20GeV]", and check the log, the simulation is all 500 MeV electrons. This may be part of the problem.
> 
> Interesting. I will try to disable download cache, in case it causes wrong steering files to be retrieved from it.

Can you confirm that I'm not hallucinating this?

**Comment by steinber** [comment]
Oh well - once the tests are restarted, I can't see the old output.  Fingers crossed and I'll check  in an hour.

**Comment by veprbl** [comment]
> Oh well - once the tests are restarted, I can't see the old output. Fingers crossed and I'll check in an hour.

No, it's all there. You can always open previous pipeline, and if jobs are re-run the previous logs are also there.
https://eicweb.phy.anl.gov/EIC/benchmarks/detector_benchmarks/-/pipelines/104131

> > > So the checks failed, but it's not an LFHCAL problem, but "ecal_gaps" raising a divide-by-zero error.
> > 
> > 
> > Correct. That might be a relevant issue.
> 
> Relevant to ECAL :)

Relevant to benchmarks that fail to get correct steering files ;)

> 
> > > Something is very weird - when I go into the "sim:lfhcal: [neutron, 20GeV]", and check the log, the simulation is all 500 MeV electrons. This may be part of the problem.
> > 
> > 
> > Interesting. I will try to disable download cache, in case it causes wrong steering files to be retrieved from it.
> 
> Can you confirm that I'm not hallucinating this?

I did, it shows e- in that ddsim debug output.

**Comment by steinber** [comment]
I just checked the LFHCAL sim jobs and the several I checked at random *did* match between the job name (neutron, 5 GeV) and the actual particle simulated.

**Comment by steinber** [comment]
Plots aren't empty anymore but there are a few things to check before I'm happy - still I think it's worth merging just to have something in place.

**Comment by steinber** [comment]
Still more to tweak, but the energy resolution plots for gamma, neutrons, pi- are not absurd anymore (but I need to have plots for the measured energy scale now too, and a quick determination of EM/HAD weights).

**Review by veprbl — APPROVED** [review]

**Comment by steinber** [comment]
Hi @veprbl and @wdconinc - a whole lot of checks failed just now, and not just in LFHCAL benchmarks (but yes one of mine failed too).  Is this something I need to address or something upstream?

**Comment by veprbl** [comment]
We somehow overfilled scratch space on eicweb. I will restart pipeline once this is resolved.

**Pull request merged**